### PR TITLE
New feature: graph join

### PR DESCRIPTION
**Pull request by GanzuraTheConsumer**
Implementing the join of two graphs, as raised in [this](https://github.com/igraph/igraph/issues/2494) issue.

### CODE REVIEW

File: examples/simple/igraph_join.c [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <https://www.gnu.org/licenses/>.
+*/
```
Comment (szhorvat) [inline examples/simple/igraph_join.c:18]: Can you please change all copyright headers in _newly added_ files to be the same as in `tests/unit/igraph_is_biconnected.c`?

File: src/operators/join.c [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+ * Both graphs need to have the same directedness, i.e. either both
+ * directed or both undirected. As of now, the join of directed graphs
+ * is unimplemented.
+ *
+ * </para><para>
+ * The current version of this function cannot handle graph, vertex
+ * and edge attributes, they will be lost.
+ *
+ * \param res  Pointer to an uninitialized graph object, the result
+ *        will stored here.
```
Comment (szhorvat) [inline src/operators/join.c:58]:
```suggestion
 *        will be stored here.
```

**Comment by szhorvat** [comment]
> I would suggest for each vertex _v_, _u_ in graphs _G1_, _G2_ we add edges (_v_, _u_), (_u_, _v_) to maintain completeness.

Alright, let's do this. Both NetworkX (`full_join()`) and Mathematica (`GraphJoin[]`) do the same.

I realize I added a review while this PR was still a draft. Let me know when it is ready to be reviewed.

File: tests/unit/igraph_join.c [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+    igraph_small(&left, 4, IGRAPH_UNDIRECTED, 0,1, 1,2, 2,2, -1);
+    igraph_small(&right, 0, IGRAPH_UNDIRECTED, -1);
+    igraph_join(&joined, &left, &right);
+    IGRAPH_ASSERT(igraph_ecount(&joined) != igraph_ecount(&left)
+                  || igraph_vcount(&joined) != igraph_vcount(&left));
+    
+    igraph_destroy(&left);
+    igraph_destroy(&right);
+    igraph_destroy(&joined);
+
```
Comment (szhorvat) [inline tests/unit/igraph_join.c:74]:
```suggestion

    VERIFY_FINALLY_STACK();

```

**Review by szhorvat — APPROVED** [review]

**Comment by szhorvat** [comment]
Thank you for implementing this. It looks good to me!

@ntamas, can you have a look as well and merge if you're happy? Do you prefer the name `igraph_join()` (consistent with existing graph operators) or `igraph_graph_join()` (more explicit name for a lesser-known graph operator)? I'm slightly leaning towards simply `igraph_join()`, i.e. the current name used in this PR.

> Should I be adding `igraph_join_many`, or is that out of scope for this PR?

I would say no, unless the concept is already in common use (to my knowledge, it is not).

**Comment by ntamas** [comment]
`igraph_join()` is the preferred version for me. I'll start a review soon.

**Review by ntamas — APPROVED** [review]

**Comment by szhorvat** [comment]
Somehow we forgot to include this in the changelog. I updated it retroactively, see https://github.com/igraph/igraph/commit/c12950899439815cb5fdf07da452844ffde8014f  @GanzuraTheConsumer, let me know if you'd like to be credited differently (real name only, GitHub name only, or both).

**Comment by GanzuraTheConsumer** [comment]
> Somehow we forgot to include this in the changelog. I updated it retroactively, see [c129508](https://github.com/igraph/igraph/commit/c12950899439815cb5fdf07da452844ffde8014f). @GanzuraTheConsumer, let me know if you'd like to be credited differently (real name only, GitHub name only, or both).

I see in the changelog I'm credited with both. That works for me!

**Pull request merged**